wicket-core/src/main/java/org/apache/wicket/pageStore/FilePageStore.java

```diff
@@ -0,0 +1,569 @@
+
+	/**
+	 * A cache that holds all page stores.
+	 */
+	private static final ConcurrentMap<String, FilePageStore> FILE_STORES = new ConcurrentHashMap<>();
+
+	private final String applicationName;
+
+	private final ISerializer serializer;
+
+	private final Bytes maxSizePerSession;
+
+	private final NestedFolders folders;
+
+	/**
+	 * Create a store that supports {@link SerializedPage}s only.
+	 * 
+	 * @param applicationName
+	 *            name of application
+	 * @param fileStoreFolder
+	 *            folder to store to
+	 * @param maxSizePerSession
+	 *            maximum size per session
+	 * 
+	 * @see SerializingPageStore
+	 */
+	public FilePageStore(String applicationName, File fileStoreFolder, Bytes maxSizePerSession)
+	{
+		this(applicationName, fileStoreFolder, maxSizePerSession, null);
+	}
+
+	/**
+	 * Create a store to disk.
+	 * 
+	 * @param applicationName
+	 *            name of application
+	 * @param fileStoreFolder
+	 *            folder to store to
+	 * @param maxSizePerSession
+	 *            maximum size per session
+	 * @param serializer
+	 *            for serialization of pages
+	 */
+	public FilePageStore(String applicationName, File fileStoreFolder, Bytes maxSizePerSession,
+		ISerializer serializer)
+	{
+		this.applicationName = Args.notNull(applicationName, "applicationName");
+		this.folders = new NestedFolders(new File(fileStoreFolder, applicationName + "-filestore"));
+		this.maxSizePerSession = Args.notNull(maxSizePerSession, "maxSizePerSession");
+		this.serializer = serializer; // optional
+
+		if (FILE_STORES.containsKey(applicationName))
+		{
+			throw new IllegalStateException(
+				"Store for application with key '" + applicationName + "' already exists.");
+		}
+		FILE_STORES.put(applicationName, this);
+	}
+
+	/**
+	 * Versioning is supported if a serializer was provided to the constructor.
+	 */
+	@Override
+	public boolean supportsVersioning()
+	{
+		return serializer != null;
+	}
+
+	@Override
+	public void destroy()
+	{
+		log.debug("Destroying...");
+
+		FILE_STORES.remove(applicationName);
+
+		log.debug("Destroyed.");
+	}
+
+	private File getPageFile(String sessionId, int id, boolean create)
+	{
+		File folder = folders.get(sessionId, create);
+
+		return new File(folder, id + FILE_SUFFIX);
+	}
+
+	@Override
+	public boolean canBeAsynchronous(IPageContext context)
+	{
+		// session attribute must be added here *before* any asynchronous calls
+		// when session is no longer available
+		getSessionAttribute(context, true);
+
+		return true;
+	}
+
+	@Override
+	public IManageablePage getPage(IPageContext context, int id)
+	{
+		SessionAttribute attribute = getSessionAttribute(context, false);
+		if (attribute == null)
+		{
+			return null;
+		}
+
+		byte[] data = readFile(attribute.identifier, id);
+		if (data == null) {
+			return null;
+		}
+		
+		IManageablePage page;
+		if (serializer == null)
+		{
+			page = new SerializedPage(id, "unknown", data);
```

Comment:
What is `unknown` here ? The session id ? What could be the consequences of using it ?